database/yugabytedb/yugabytedb.go

```diff
@@ -363,50 +405,55 @@ func (c *YugabyteDB) ensureLockTable() error {
 	return nil
 }
 
-func ExecuteInTx(ctx context.Context, db *sql.DB, opts *sql.TxOptions, fn func(tx *sql.Tx) error) (err error) {
-	// It looks like currently, YugabyteDB doesn't return an error on savepoint release.
-	// So, we have to retry the whole transaction.
-	// TODO: do more research on savepoints support, see https://github.com/yugabyte/yugabyte-db/issues/9219
+func (c *YugabyteDB) doTxWithRetry(
+	ctx context.Context,
+	txOpts *sql.TxOptions,
+	fn func(tx *sql.Tx) error,
+) error {
+	backOff := c.newBackoff(ctx)
 
-	const maxRetries = 50
-	retryCount := 0
-
-	for {
-		if retryCount-1 >= maxRetries {
-			return fmt.Errorf("%w: reached %d attempts", ErrMaxRetriesExceeded, retryCount-1)
-		}
-
-		tx, err := db.BeginTx(ctx, opts)
+	return backoff.Retry(func() error {
+		tx, err := c.db.BeginTx(ctx, txOpts)
 		if err != nil {
-			return err
+			return backoff.Permanent(err)
 		}
 
-		if err := fn(tx); err != nil {
-			_ = tx.Rollback()
-
-			if !errIsRetryable(err) {
-				return err
-			}
+		defer tx.Rollback()
 
-			retryCount++
+		if err := fn(tx); err != nil && !errIsRetryable(err) {
+			return backoff.Permanent(err)
+		} else if err != nil {
+			return err
+		}
 
-			continue
+		if err := tx.Commit(); err != nil && !errIsRetryable(err) {
+			return backoff.Permanent(err)
+		} else if err != nil {
+			return err
 		}
 
-		if err := tx.Commit(); err != nil {
-			_ = tx.Rollback()
+		return nil
+	}, backOff)
+}
 
-			if !errIsRetryable(err) {
-				return err
-			}
+func (c *YugabyteDB) newBackoff(ctx context.Context) backoff.BackOff {
```

Comment:
Shouldn't this return a `*backoff.BackOff`?

Reply:
`backoff.Retry` accepts `backoff.Backoff` by value. So IMO there is no point in returning a pointer, because we'll have to dereference it later anyway